ceph-mon/lib/charms/ceph_mon/v0/ceph_cos_agent.py

```diff
@@ -45,25 +45,41 @@ def __init__(self, charm):
 
     def _on_refresh(self, event):
         """Enable prometheus on relation change"""
-        if not ceph_utils.is_bootstrapped():
-            logger.debug("not bootstrapped, defer _on_refresh: %s", event)
-            event.defer()
+        if not self._charm.unit.is_leader():
+            logger.debug("Not the charm leader, skipping refresh cb.")
             return
-        logger.debug("refreshing cos_agent relation")
-        if self._charm.unit.is_leader():
-            self.mgr_config_set_rbd_stats_pools()
+
+        if hasattr(self._charm, "_cos_agent_refresh_cb") and callable(self._charm._cos_agent_refresh_cb):
+            self._charm._cos_agent_refresh_cb(event) 
+        else:
+            # ceph mon failback
+            if not ceph_utils.is_bootstrapped():
+                logger.debug("not bootstrapped, defer _on_refresh: %s", event)
+                event.defer()
+                return
+
+            logger.debug("refreshing cos_agent relation")
             ceph_utils.mgr_enable_module("prometheus")
+
+        self.mgr_config_set_rbd_stats_pools()
         super()._on_refresh(event)
 
     def _on_relation_departed(self, event):
         """Disable prometheus on depart of relation"""
-        if self._charm.unit.is_leader() and ceph_utils.is_bootstrapped():
-            logger.debug(
-                "is_leader and is_bootstrapped, running rel departed: %s",
-                event,
-            )
-            ceph_utils.mgr_disable_module("prometheus")
-            logger.debug("module_disabled")
+        if self._charm.unit.is_leader():
+            logger.debug("Not the charm leader, skipping relation_departed: %s.", event)
+
+        if hasattr(self._charm, "_cos_agent_departed_cb") and callable(self._charm._cos_agent_departed_cb):
+            self._charm._cos_agent_departed_cb(event)
```

Comment:
Same as above wrt to callbacks on the client class, suggest to inject this callback in the constructor

Reply:
ack